crates/turbopack-ecmascript/src/tree_shake/mod.rs

```diff
@@ -0,0 +1,383 @@
+}
+
+#[turbo_tasks::value(shared, serialization = "none", eq = "manual")]
+pub(crate) enum SplitResult {
+    Ok {
+        /// `u32` is a index to `modules`.
+        #[turbo_tasks(debug_ignore, trace_ignore)]
+        entrypoints: FxHashMap<Key, u32>,
+
+        #[turbo_tasks(debug_ignore, trace_ignore)]
+        modules: Vec<ParseResultVc>,
+
+        #[turbo_tasks(debug_ignore, trace_ignore)]
+        deps: FxHashMap<u32, Vec<u32>>,
+    },
+    Unparseable,
+    NotFound,
+}
+
+impl PartialEq for SplitResult {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Ok { .. }, Self::Ok { .. }) => false,
+            _ => core::mem::discriminant(self) == core::mem::discriminant(other),
+        }
+    }
+}
+
+#[turbo_tasks::function]
+pub(super) fn split_module(asset: EcmascriptModuleAssetVc) -> SplitResultVc {
+    split(asset.origin_path(), asset.parse())
+}
+
+#[turbo_tasks::function]
+pub(super) async fn split(path: FileSystemPathVc, parsed: ParseResultVc) -> Result<SplitResultVc> {
+    let filename = path.await?.file_name().to_string();
+    let parse_result = parsed.await?;
+
+    match &*parse_result {
+        ParseResult::Ok {
+            program: Program::Module(module),
+            comments,
+            eval_context,
+            source_map,
+            globals,
+            ..
+        } => {
+            let (mut dep_graph, items) = Analyzer::analyze(module);
+
+            dep_graph.handle_weak(Mode::Production);
+
+            let SplitModuleResult {
+                entrypoints,
+                part_deps,
+                modules,
+            } = dep_graph.split_module(&format!("./{filename}").into(), &items);
+
+            let modules = modules
+                .into_iter()
+                .map(|module| {
+                    let program = Program::Module(module);
+                    let eval_context = EvalContext::new(&program, eval_context.unresolved_mark);
+
+                    ParseResultVc::cell(ParseResult::Ok {
+                        program,
+                        globals: globals.clone(),
+                        comments: comments.clone(),
+                        source_map: source_map.clone(),
+                        eval_context,
+                    })
+                })
+                .collect();
+
+            Ok(SplitResult::Ok {
+                entrypoints,
+                deps: part_deps,
+                modules,
+            }
+            .cell())
+        }
+        ParseResult::NotFound => Ok(SplitResult::NotFound.cell()),
+        _ => Ok(SplitResult::Unparseable.cell()),
+    }
+}
+
+#[turbo_tasks::function]
+pub(super) async fn part_of_module(
+    split_data: SplitResultVc,
+    part: Option<ModulePartVc>,
+) -> Result<ParseResultVc> {
+    let split_data = split_data.await?;
+
+    let part_id = match part {
+        Some(part) => get_part_id(&split_data, part).await?,
+        None => bail!("part {:?} is not found in the module", part),
```

Comment:
When it's not valid to call `part_of_module` with `part = None` why does the signature allow that?